File: swift/Sources/DopamineCore/Sdf.swift
// Baked-SDF blob decoder — the Swift port of `engine/sdf.ts`'s `decodeSdf`.
//
// PORTABLE (no `#if canImport(Metal)` guard): pure byte decoding of the inline
// `data:` SDF blob the build step inlines into the `.dope`. The native runner
// (MetalPassRunner) uploads the decoded single-channel bytes as an R8 texture;
// this decode step is shared and Linux-buildable, mirroring the web posture
// (decode at runtime, never re-bake).
//
// Encoding (from the baker, `engine/sdf.ts`): a 4-byte header — magic 'D','S'
// (0x44, 0x53), then size hi/lo — followed by `size*size` single-channel 0..255
// distance bytes, the whole thing base64-encoded behind a
// `data:application/octet-stream;base64,` URI.

import Foundation

/// A decoded SDF ready to upload: the single-channel distance bytes + the square
/// grid `size`. Mirror of the web `DecodedSdf` (minus the range/viewBox metadata,
/// which the native runtimes read off `render.pass` separately).
public struct DopeDecodedSdf {
    public let size: Int
    /// `size`×`size` single-channel (0..255) distance bytes.
    public let bytes: [UInt8]
    public init(size: Int, bytes: [UInt8]) {
        self.size = size
        self.bytes = bytes
    }
}

private let sdfDataPrefix = "data:application/octet-stream;base64,"
private let sdfMagic0: UInt8 = 0x44  // 'D'
private let sdfMagic1: UInt8 = 0x53  // 'S'

/// Decode a baked SDF `data:` URI (or a bare base64 blob) back to its raw
/// single-channel distance bytes. Validates the magic + the declared size
/// against the byte count; returns nil on a bad magic / size mismatch / invalid
/// base64 (the caller degrades to the analytic fallback rather than crashing).
public func decodeDopeSdf(_ dataURI: String) -> DopeDecodedSdf? {
    let b64 = dataURI.hasPrefix(sdfDataPrefix)
        ? String(dataURI.dropFirst(sdfDataPrefix.count))
        : dataURI  // tolerate a bare base64 blob too
    guard let blob = Data(base64Encoded: b64) else { return nil }
    guard blob.count >= 4, blob[0] == sdfMagic0, blob[1] == sdfMagic1 else { return nil }
    let size = (Int(blob[2]) << 8) | Int(blob[3])
    let body = blob.dropFirst(4)
    guard body.count == size * size else { return nil }
    return DopeDecodedSdf(size: size, bytes: [UInt8](body))
}

File: android/dopamine-core/src/main/kotlin/ai/dopamine/core/Sdf.kt
// Baked-SDF runtime DECODER — the pure-JVM port of `engine/sdf.ts`'s `decodeSdf`.
//
// An effect's icon outline is baked at build time into a tiny, self-contained
// distance field and inlined into the `.dope` as a `data:` base64 blob (see
// `engine/sdf.ts` for the baker + the wire format). At RUNTIME the platform only
// DECODES + SAMPLES it (never re-bakes), so this is all the native side needs.
//
// This lives in `dopamine-core` (PURE JVM, no `android.*`) so the decode is
// testable on the parity grid with no Android SDK; the GL upload/bind of the
// decoded bytes lives in `dopamine-gl`.
//
// Wire format (must match `engine/sdf.ts`):
//   - a 4-byte header: magic 'D'(0x44), 'S'(0x53), then size hi/lo (big-endian),
//   - followed by size×size single-channel (0..255) distance bytes,
//   - the whole blob base64-encoded behind a `data:...;base64,` URI.

package ai.dopamine.core

import java.util.Base64

/** A decoded SDF ready to upload: the grid `size` + its `size*size` distance bytes. */
class DopeDecodedSdf(val size: Int, val bytes: ByteArray)

/** The MIME prefix the baker writes (`engine/sdf.ts`'s `SDF_DATA_PREFIX`). */
private const val SDF_DATA_PREFIX = "data:application/octet-stream;base64,"

private const val MAGIC0 = 0x44 // 'D'
private const val MAGIC1 = 0x53 // 'S'

/**
 * Decode a baked-SDF `data:...;base64,` (or bare base64) blob → its raw
 * single-channel distance bytes + grid size. Validates the magic + that the
 * declared size matches the byte count; returns null on a bad/short/mismatched
 * blob (so a caller can fall back to the analytic path rather than throwing).
 *
 * Bytes are SIGNED in Kotlin — the size header is read with `.toInt() and 0xFF`.
 */
fun decodeDopeSdf(dataURI: String): DopeDecodedSdf? {
    val b64 = if (dataURI.startsWith(SDF_DATA_PREFIX)) dataURI.substring(SDF_DATA_PREFIX.length) else dataURI
    val blob = try {
        Base64.getDecoder().decode(b64.trim())
    } catch (e: IllegalArgumentException) {
        return null
    }
    if (blob.size < 4) return null
    if ((blob[0].toInt() and 0xFF) != MAGIC0 || (blob[1].toInt() and 0xFF) != MAGIC1) return null
    val size = ((blob[2].toInt() and 0xFF) shl 8) or (blob[3].toInt() and 0xFF)
    val bytes = blob.copyOfRange(4, blob.size)
    if (bytes.size != size * size) return null
    return DopeDecodedSdf(size, bytes)
}
